File: csharp-selenium-crawler/Pages/ChallengingDomPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using OpenQA.Selenium;

namespace Crawler.Pages
{
    public class ChallengingDomPage
    {
        private IWebDriver _driver;
        private const string Url = "https://the-internet.herokuapp.com/challenging_dom";

        public ChallengingDomPage(IWebDriver driver)
        {
            _driver = driver;
        }

        public void Open()
        {
            _driver.Navigate().GoToUrl(Url);
        }

        public string GetUrl()
        {
            return Url;
        }

        public List<LinkInfo> GetFeatureLinks()
        {
            var links = _driver.FindElements(By.TagName("a"));
            var featureLinks = new List<LinkInfo>();

            foreach (var link in links)
            {
                var name = link.Text.Trim();
                var href = link.GetAttribute("href");

                if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(href))
                {
                    featureLinks.Add(new LinkInfo { Name = name, Url = href });
                }
            }

            return featureLinks;
        }
    }

    public class LinkInfo
    {
        public string Name { get; set; } = "";
        public string Url { get; set; } = "";
    }
}

File: java-selenium-crawler/src/main/java/com/crawler/pages/ChallengingDomPage.java
package com.crawler.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

public class ChallengingDomPage {
    private WebDriver driver;
    private final String url = "https://the-internet.herokuapp.com/challenging_dom";

    public ChallengingDomPage(WebDriver driver) {
        this.driver = driver;
    }

    public void open() {
        driver.get(url);
    }

    public String getUrl() {
        return url;
    }

    public List<Map<String, String>> getFeatureLinks() {
        List<Map<String, String>> featureLinks = new ArrayList<>();
        List<WebElement> links = driver.findElements(By.tagName("a"));

        for (WebElement link : links) {
            String name = link.getText().trim();
            String href = link.getAttribute("href");

            if (!name.isEmpty() && href != null) {
                Map<String, String> linkData = new HashMap<>();
                linkData.put("name", name);
                linkData.put("url", href);
                featureLinks.add(linkData);
            }
        }
        return featureLinks;
    }
}
